File: refractor-agent/scripts/embed.py
#!/usr/bin/env python3
"""Build the refractor vector library from the editable dicts.

Scans ``dicts/*.yml`` (skipping ``schema.md``), validates each dictionary, embeds
every refraction entry, and (idempotently) rebuilds the LanceDB bucket.

Exit non-zero on dictionary validation failure (e.g. duplicate ``(pattern,color)``).
"""

from __future__ import annotations

import argparse
import sys
from pathlib import Path

import yaml

from refract_store import Embedder, Store, load_env

DICT_GLOB = "dicts/*.yml"
ENUM_FILE = "enum.yml"


def load_dicts(dict_dir: Path) -> list[dict]:
    dicts = []
    for path in sorted(dict_dir.glob("*.yml")):
        with path.open(encoding="utf-8") as fh:
            doc = yaml.safe_load(fh)
        if not isinstance(doc, dict) or "refractions" not in doc:
            continue
        doc["_file"] = str(path)
        dicts.append(doc)
    return dicts


def load_enum(dict_dir: Path) -> dict[str, list[str]]:
    """Load the controlled pattern/color enum (dicts/enum.yml)."""
    path = dict_dir / ENUM_FILE
    with path.open(encoding="utf-8") as fh:
        doc = yaml.safe_load(fh)
    if not isinstance(doc, dict) or "pattern" not in doc or "color" not in doc:
        raise SystemExit(f"{path} must define pattern and color lists")
    return doc


def validate(doc: dict, enum: dict[str, list[str]]) -> list[str]:
    """Return a list of validation errors (empty means valid)."""
    errors = []
    for key in ("brand", "series", "year", "refractions"):
        if key not in doc:
            errors.append(f"[{doc.get('_file')}] missing top-level '{key}'")
    seen = {}
    for i, refr in enumerate(doc.get("refractions", [])):
        pat = refr.get("pattern")
        color = refr.get("color")
        key = (pat, color)
        if key in seen:
            errors.append(
                f"[{doc.get('_file')}] duplicate (pattern,color)=({pat},{color}) "
                f"between #{seen[key]} and #{i}; same pattern+color must be one entry"
            )
        seen[key] = i
        for field in ("name", "name_en", "pattern", "color", "keywords"):
            if field not in refr:
                errors.append(f"[{doc.get('_file')}] refraction #{i} missing '{field}'")
        if not isinstance(refr.get("keywords"), list):
            errors.append(f"[{doc.get('_file')}] refraction #{i} 'keywords' must be a list")
        if refr.get("pattern") not in enum["pattern"]:
            errors.append(
                f"[{doc.get('_file')}] refraction #{i} pattern={refr.get('pattern')!r} "
                f"not in controlled enum (see dicts/enum.yml)"
            )
        if refr.get("color") not in enum["color"]:
            errors.append(
                f"[{doc.get('_file')}] refraction #{i} color={refr.get('color')!r} "
                f"not in controlled enum (see dicts/enum.yml)"
            )
    return errors


def main() -> int:
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--dict-dir", default=str(Path(__file__).parent.parent / "dicts"))
    ap.add_argument("--db", default=str(Path(__file__).parent.parent / "db" / "refractors.lance"))
    args = ap.parse_args()

    load_env()
    dict_dir = Path(args.dict_dir)
    dicts = load_dicts(dict_dir)
    if not dicts:
        print(f"no dicts found under {dict_dir}", file=sys.stderr)
        return 1

    enum = load_enum(dict_dir)
    all_errors: list[str] = []
    for doc in dicts:
        all_errors.extend(validate(doc, enum))
    if all_errors:
        for e in all_errors:
            print(f"VALIDATION ERROR: {e}", file=sys.stderr)
        return 1

    embedder = Embedder()
    print("embedder: " + ("remote" if embedder.remote else "local-fallback (set EMBED_*)"))
    if embedder.remote:
        # remote dim unknown up front; fine, LanceDB infers from first row
        pass

    rows = []
    for doc in dicts:
        brand = doc["brand"]
        series = doc["series"]
        year = doc["year"]
        for refr in doc["refractions"]:
            text = " ".join(
                [refr["name"], refr["name_en"], *refr["keywords"], refr["pattern"], refr["color"]]
            )
            rows.append(
                {
                    "id": f"{brand}-{series}-{refr['pattern']}-{refr['color']}",
                    "brand": brand,
                    "series": series,
                    "year": year,
                    "name": refr["name"],
                    "name_en": refr["name_en"],
                    "pattern": refr["pattern"],
                    "color": refr["color"],
                    "keywords": refr["keywords"],
                    "text": text,
                }
            )

    vecs = embedder.embed([r["text"] for r in rows])
    for r, v in zip(rows, vecs):
        r["vector"] = v

    Store(args.db).reset(rows)
    print(f"wrote {len(rows)} refraction entries -> {args.db}")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: refractor-agent/node/src/embed.ts
/** Build the refractor vector library from the single editable dict.
 *
 * TypeScript port of `scripts/embed.py` (CLI included). Reads
 * `dicts/refractions.yml` (one entry per database row), validates it, and
 * idempotently rebuilds the vector store (Supabase pgvector by default, lance
 * fallback). Each entry's `pattern` / `color` / `keywords` are
 * employee-edited; `id` / `text` / `vector` are generated here.
 */

import { join } from "node:path";

import { AGENT_ROOT, loadEnv } from "./env.ts";
import {
  DICT_FILE,
  buildNameRows,
  buildRows,
  loadDictionary,
  validateDictionary,
} from "./dict.ts";
import { BaseStore, Embedder, createStore, writeFingerprint } from "./store.ts";

export interface EmbedResult {
  rows: number;
  names: number;
}

/** Validate the dict, embed every row, and idempotently rebuild both tables. */
export async function embedDictIntoStore(
  dictDir: string,
  store: BaseStore,
  embedder: Embedder,
): Promise<EmbedResult> {
  const path = join(dictDir, DICT_FILE);
  const doc = await loadDictionary(path);
  const errors = validateDictionary(doc, path);
  if (errors.length > 0) {
    throw new Error(`dictionary validation failed:\n${errors.join("\n")}`);
  }

  const rows = buildRows(doc);
  const vectors = await embedder.embed(rows.map((row) => row.text));
  for (const [index, vector] of vectors.entries()) {
    rows[index].vector = vector;
  }
  const names = buildNameRows(doc);

  await store.reset(rows);
  await store.resetNames(names);
  return { rows: rows.length, names: names.length };
}

/** Fresh-store rebuild used by the dict→store self-healing sync. */
export async function rebuildStore(dictDir: string, dbPath?: string): Promise<void> {
  const store = await createStore(dbPath);
  await embedDictIntoStore(dictDir, store, new Embedder());
}

async function main(): Promise<number> {
  loadEnv();
  let dictDir = join(AGENT_ROOT, "dicts");
  let dbPath: string | undefined;
  const argv = process.argv.slice(2);
  for (let i = 0; i < argv.length; i++) {
    if (argv[i] === "--dict-dir") dictDir = argv[++i] as string;
    else if (argv[i] === "--db") dbPath = argv[++i] as string;
  }

  const embedder = new Embedder();
  console.log(`embedder: ${embedder.remote ? "remote" : "local-fallback (set EMBED_*)"}`);
  const store = await createStore(dbPath);
  const result = await embedDictIntoStore(dictDir, store, embedder);
  await writeFingerprint(dictDir);
  console.log(
    `wrote ${result.rows} refraction entries + ${result.names} series names -> ${store.constructor.name}`,
  );
  return 0;
}

if (process.argv[1] !== undefined && import.meta.url === new URL(`file://${process.argv[1]}`).href) {
  main().then(
    (code) => process.exit(code),
    (error) => {
      console.error(error instanceof Error ? error.message : error);
      process.exit(1);
    },
  );
}
